src/gurt/memory.c

```diff
@@ -0,0 +1,217 @@
+
+#define MM_MAGIC	0xa5a55a5a
+#define MM_MIN_SIZE	0x80	/* 128 bytes */
+#define MM_MAX_NBUCKETS	10
+
+struct mm_item {
+	d_list_t	 mi_link;
+	uint32_t	 mi_bucket;
+	uint32_t	 mi_magic;
+	char		 mi_buff[0];
+};
+
+struct mm_bucket {
+	d_list_t	   mb_head;
+	pthread_spinlock_t mb_lock;
+	uint32_t	   mb_count;
+};
+
+struct mm_manager {
+	struct mm_bucket *mm_buckets;
+	uint32_t	  mm_count;
+};
+
+static struct mm_manager dmm;
+
+int
+d_mm_init(size_t n)
+{
+	struct mm_bucket *bucket;
+	size_t i;
+	int rc;
+
+	if (unlikely(n < 2 || n >= MM_MAX_NBUCKETS))
+		D_GOTO(out, rc = -DER_INVAL);
+
+	if (unlikely(dmm.mm_count != 0))
+		D_GOTO(out, rc = -DER_ALREADY);
+
+	dmm.mm_buckets = malloc(n * sizeof(*dmm.mm_buckets));
+	if (unlikely(dmm.mm_buckets == NULL))
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	for (i = 0; i < n; i++) {
+		bucket = &dmm.mm_buckets[i];
+		bucket->mb_count = 0;
+		D_INIT_LIST_HEAD(&bucket->mb_head);
+		rc = D_SPIN_INIT(&bucket->mb_lock, PTHREAD_PROCESS_PRIVATE);
+		if (unlikely(rc))
+			D_GOTO(out_free, rc);
+	}
+	dmm.mm_count = n;
+	D_GOTO(out, rc = 0);
+
+out_free:
+	while (i > 0) {
+		bucket = &dmm.mm_buckets[--i];
+		D_SPIN_DESTROY(&bucket->mb_lock);
+	}
+	free(dmm.mm_buckets);
+out:
+	return rc;
+}
+
+void
+d_mm_fini(void)
+{
+	struct mm_bucket *bucket;
+	size_t i, count = dmm.mm_count;
+
+	if (unlikely(count == 0))
+		return;
+
+	d_mm_flush();
+	dmm.mm_count = 0;
+	for (i = 0; i < count; i++) {
+		bucket = &dmm.mm_buckets[i];
+		D_ASSERT(bucket->mb_count == 0);
+		D_SPIN_DESTROY(&bucket->mb_lock);
+	}
+	free(dmm.mm_buckets);
+}
+
+void *
+d_mm_alloc(size_t size)
+{
+	struct mm_bucket *bucket;
+	struct mm_item *item = NULL;
+	void *ptr;
+	size_t i, nsize;
+
+	size += sizeof(struct mm_item);
+	for (i = 0, nsize = MM_MIN_SIZE; nsize < size; i++)
+		nsize <<= 1;
+
+	if (likely(i < dmm.mm_count)) {
+		bucket = &dmm.mm_buckets[i];
+		D_SPIN_LOCK(&bucket->mb_lock);
+		item = d_list_pop_entry(&bucket->mb_head,
+					struct mm_item, mi_link);
+		if (likely(item != NULL))
+			bucket->mb_count--;
+		D_SPIN_UNLOCK(&bucket->mb_lock);
+	}
+
+	if (unlikely(item == NULL)) {
+		item = malloc(nsize);
+		if (unlikely(item == NULL))
+			return NULL;
+	}
+
+	item->mi_magic  = MM_MAGIC;
+	item->mi_bucket = i;
+
+	ptr = &item->mi_buff[0];
+	memset(ptr, 0, size - sizeof(struct mm_item));
+
+	return ptr;
+}
+
+void *
+d_mm_realloc(void *ptr, size_t size)
+{
+	struct mm_item *item = container_of(ptr, struct mm_item, mi_buff);
+	void *ptr2;
+
+	if (unlikely(ptr == NULL))
+		return d_mm_alloc(size);
+
+	if (unlikely(item->mi_magic != MM_MAGIC))
+		return realloc(ptr, size);
+
+	ptr2 = d_mm_alloc(size);
+	if (unlikely(ptr2 == NULL))
+		return realloc(ptr, size);
+
+	memmove(ptr2, ptr, size);
+
+	d_mm_free(ptr);
+	return ptr2;
+}
+
+void
+d_mm_free(void *ptr)
+{
+	struct mm_bucket *bucket;
+	struct mm_item *item = container_of(ptr, struct mm_item, mi_buff);
+
+	if (unlikely(ptr == NULL))
+		return;
+
+	if (unlikely(item->mi_magic != MM_MAGIC)) {
```

Comment:
You can't do this, mixing and matching of allocators does not and can not work.

Reply:
This is just sanity check for not pared call. It works only if this function free system allocated memory. We have a mess in DAOS memory allocations. I encountered a lot of places where memory allocated by D_ALLOC() but then passwd to DRPC functions and free by system free() there. Unfortunately I cannot workaround this.

Reply:
DRPC should all be using the macros now, I'm not aware of any places which are not in that area.

We proactively test for, and resolve places where D_ALLOC/D_FREE are mixed with alloc/free but it's not entirely consistent and I suspect it never will be, or not to the level of reliability where we are able to change D_ALLOC to be based on something else.

I've got (another) PR to improve the macros at https://github.com/daos-stack/daos/pull/3566 but it's an ongoing work as you've seen.

Reply:
Sorry, I was wrong.  DRPC in DAOS should all be using the allocator which is a wrapper around D_ALLOC, however I didn't update all the test code to use this.

Reply:
It works in one way only. d_mm_free() can free memory allocated by malloc().

Reply:
It doesn't matter that it only works one way, you still cannot do this.

What you're doing is taking a buffer that was allocated by malloc(), taking an backwards offset from the pointer and looking for a magic value, then freeing the pointer if the value doesn't mix.  This is against the C standard, prone to failure and occasional segaults and will break many compliance tools.

Reply:
Ok. Change to ASSERT.